Prevent the TE fuser from getting datatypes it can't handle

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44152 Fix frac in CUDA fuser
* **#44160 Prevent the TE fuser from getting datatypes it can't handle**
* #44157 [te] Fix casting of unsigned char, and abs(int)

Differential Revision: [D23528508](https://our.internmc.facebook.com/intern/diff/D23528508)

bertmaher committed:
Prevent the TE fuser from getting datatypes it can't handle

[ghstack-poisoned]

bertmaher committed:
Update on "Prevent the TE fuser from getting datatypes it can't handle"




[ghstack-poisoned]

bertmaher committed:
Update on "Prevent the TE fuser from getting datatypes it can't handle"




[ghstack-poisoned]

ZolotukhinM commented on test/test_jit_fuser_te.py:
```diff
@@ -1210,7 +1213,8 @@ def rand(dtype, device="cuda"):
             t = torch.jit.trace(fn, (x,))
             self.assertEqual(ref, t(x))
-            self.assertAllFused(t.graph_for(x))
+            if dtype in supported_dtypes:
```
What do we test when we run scripted but not fused graphs (when the dtype is unsupported)?

bertmaher replied:
Basically this is just testing that we don't generate something that throws (which is what would happen before this diff).  But it's kind of ridiculous to test that for every single unary op.  Let me split the unsupported dtypes into a separate tiny test.

ZolotukhinM commented on test/test_jit_fuser_te.py:
```diff
@@ -1146,24 +1146,27 @@ def rand(dtype, device="cuda"):
             torch.int32,
             torch.int64,
             # torch.float16,
-            # torch.bfloat16,
```
Is it intentional that `torch.float16` remains commented out?

bertmaher replied:
Yeah, some fixes for float16 are in the works.  This should probably be a TODO but I expect it to land in a day or so.

ZolotukhinM reviewed (approved):
👍

bertmaher committed:
Update on "Prevent the TE fuser from getting datatypes it can't handle"




[ghstack-poisoned]

bertmaher committed:
Update on "Prevent the TE fuser from getting datatypes it can't handle"


Differential Revision: [D23528508](https://our.internmc.facebook.com/intern/diff/D23528508)

[ghstack-poisoned]

bertmaher committed:
Update on "Prevent the TE fuser from getting datatypes it can't handle"


Differential Revision: [D23528508](https://our.internmc.facebook.com/intern/diff/D23528508)

[ghstack-poisoned]

bertmaher committed:
Update on "Prevent the TE fuser from getting datatypes it can't handle"


Differential Revision: [D23528508](https://our.internmc.facebook.com/intern/diff/D23528508)

[ghstack-poisoned]

bertmaher committed:
Update on "Prevent the TE fuser from getting datatypes it can't handle"


Differential Revision: [D23528508](https://our.internmc.facebook.com/intern/diff/D23528508)

[ghstack-poisoned]

bertmaher committed:
Update on "Prevent the TE fuser from getting datatypes it can't handle"


Differential Revision: [D23528508](https://our.internmc.facebook.com/intern/diff/D23528508)

[ghstack-poisoned]
